exchange/exchange_test.go

```diff
@@ -274,6 +274,193 @@ func TestDebugBehaviour(t *testing.T) {
 	}
 }
 
+func TestReturnCreativeEndToEnd(t *testing.T) {
+	sampleAd := "<?xml version=\"1.0\" encoding=\"UTF-8\"?><VAST ...></VAST>"
+
+	// Define test cases
+	type aTest struct {
+		desc   string
+		inExt  json.RawMessage
+		outAdM string
+	}
+	testGroups := []struct {
+		groupDesc   string
+		testCases   []aTest
+		expectError bool
+	}{
+		{
+			groupDesc: "Invalid or malformed bidRequest Ext, expect error in these scenarios",
+			testCases: []aTest{
+				{
+					desc:  "Malformed ext in bidRequest",
+					inExt: json.RawMessage(`malformed`),
+				},
+				{
+					desc:  "empty cache field",
+					inExt: json.RawMessage(`{"prebid":{"cache":{}}}`),
+				},
+			},
+			expectError: true,
+		},
+		{
+			groupDesc: "Valid bidRequest Ext but no returnCreative value specified, default to returning creative",
+			testCases: []aTest{
+				{
+					"Nil ext in bidRequest",
+					nil,
+					sampleAd,
+				},
+				{
+					"empty ext",
+					json.RawMessage(``),
+					sampleAd,
+				},
+				{
+					"bids doesn't come with returnCreative value",
+					json.RawMessage(`{"prebid":{"cache":{"bids":{}}}}`),
+					sampleAd,
+				},
+				{
+					"vast doesn't come with returnCreative value",
+					json.RawMessage(`{"prebid":{"cache":{"vastXml":{}}}}`),
+					sampleAd,
+				},
+			},
+		},
+		{
+			groupDesc: "Bids field comes with returnCreative value",
+			testCases: []aTest{
+				{
+					"Bids returnCreative set to true, return ad markup in response",
+					json.RawMessage(`{"prebid":{"cache":{"bids":{"returnCreative":true}}}}`),
+					sampleAd,
+				},
+				{
+					"Bids returnCreative set to false, don't return ad markup in response",
+					json.RawMessage(`{"prebid":{"cache":{"bids":{"returnCreative":false}}}}`),
+					"",
+				},
+			},
+		},
+		{
+			groupDesc: "Vast field comes with returnCreative value",
+			testCases: []aTest{
+				{
+					"Vast returnCreative set to true, return ad markup in response",
+					json.RawMessage(`{"prebid":{"cache":{"vastXml":{"returnCreative":true}}}}`),
+					sampleAd,
+				},
+				{
+					"Vast returnCreative set to false, don't return ad markup in response",
+					json.RawMessage(`{"prebid":{"cache":{"vastXml":{"returnCreative":false}}}}`),
+					"",
+				},
+			},
+		},
+		{
+			groupDesc: "Both Bids and Vast come with their own returnCreative value",
+			testCases: []aTest{
+				{
+					"Both false, expect empty AdM",
+					json.RawMessage(`{"prebid":{"cache":{"bids":{"returnCreative":false},"vastXml":{"returnCreative":false}}}}`),
+					"",
+				},
+				{
+					"Bids returnCreative is true, expect valid AdM",
+					json.RawMessage(`{"prebid":{"cache":{"bids":{"returnCreative":false},"vastXml":{"returnCreative":true}}}}`),
```

Comment:
Typo here as this json message doesn't match the description and is the same as the next test case.
This should probably be `{"bids":{"returnCreative":true},"vastXml":{"returnCreative":false}}`